RawSpeed/TiffEntryBE.cpp

```diff
@@ -84,67 +81,77 @@ TiffEntryBE::TiffEntryBE( TiffTag tag, TiffDataType type, uint32 count, const uc
+
+  if (num*2+1 >= bytesize)
+    ThrowTPE("TIFF, getSShort: Trying to read out of bounds");
+
+  return (short16) get2LE(data, num*2);
 }
 
-const ushort16* TiffEntryBE::getShortArray() {
-  if (!(type == TIFF_SHORT || type == TIFF_UNDEFINED))
-    ThrowTPE("TIFF, getShortArray: Wrong type 0x%x encountered. Expected Short", type);
+uint32 TiffEntryBE::getInt(uint32 num) {
+  if (type == TIFF_SHORT) return getShort(num);
+  if (!(type == TIFF_LONG || type == TIFF_OFFSET || type == TIFF_BYTE || type == TIFF_UNDEFINED || type == TIFF_RATIONAL || type == TIFF_SRATIONAL))
+    ThrowTPE("TIFF, getInt: Wrong type %u encountered. Expected Long, Offset or Undefined on 0x%x", type, tag);
 
-  if (own_data)
-    return (unsigned short*)own_data;
+  if (num*4+3 >= bytesize)
+    ThrowTPE("TIFF, getInt: Trying to read out of bounds");
 
-  own_data = new uchar8[count*2];
-  ushort16* d = (ushort16*)own_data;
-  for (uint32 i = 0; i < count; i++) {
-    d[i] = (ushort16)data[i*2+0] << 8 | (ushort16)data[i*2+1];
-  }
-  return d;
+  return get4BE(data, num*4);
 }
 
-const short16* TiffEntryBE::getSignedShortArray() {
-  if (!(type == TIFF_SSHORT))
-    ThrowTPE("TIFF, getShortArray: Wrong type 0x%x encountered. Expected SShort", type);
+int32 TiffEntryBE::getSInt(uint32 num) {
+  if (type == TIFF_SSHORT) return getSShort(num);
+  if (!(type == TIFF_SLONG || type == TIFF_UNDEFINED))
+    ThrowTPE("TIFF, getSInt: Wrong type %u encountered. Expected SLong or Undefined on 0x%x", type, tag);
+
+  if (num*4+3 >= bytesize)
+    ThrowTPE("TIFF, getSInt: Trying to read out of bounds");
 
-  if (own_data)
-    return (short16 *)own_data;
+  return get4BE(data, num*4);
+}
 
-  own_data = new uchar8[count*2];
-  ushort16* d = (ushort16*)own_data;
-  for (uint32 i = 0; i < count; i++) {
-    d[i] = (ushort16)data[i*2+0] << 8 | (ushort16)data[i*2+1];
+float TiffEntryBE::getFloat(uint32 num) {
+  if (!isFloat())
+    ThrowTPE("TIFF, getFloat: Wrong type 0x%x encountered. Expected Float or something convertible on 0x%x", type, tag);
+
+  if (type == TIFF_DOUBLE) {
+    if (num*8+7 >= bytesize)
+      ThrowTPE("TIFF, getFloat: Trying to read out of bounds");
+    return (float) get8BE(data, num*8);
+  } else if (type == TIFF_FLOAT) {
+    if (num*4+3 >= bytesize)
+      ThrowTPE("TIFF, getFloat: Trying to read out of bounds");
+    return (float) get4BE(data, num*4);
+  } else if (type == TIFF_LONG || type == TIFF_SHORT) {
+    return (float)getInt(num);
+  } else if (type == TIFF_SLONG || type == TIFF_SSHORT) {
+    return (float)getSInt(num);
+  } else if (type == TIFF_RATIONAL) {
+    uint32 a = getInt(num*2);
+    uint32 b = getInt(num*2+1);
+    if (b)
+      return (float) a/b;
+  } else if (type == TIFF_SRATIONAL) {
+    int a = (int) getInt(num*2);
+    int b = (int) getInt(num*2+1);
+    if (b)
+      return (float) a/b;
   }
-  return (short16 *)d;
+  return 0.0f;
```

Comment:
AFAICT this return "0", if the type isn't known.

Shouldn't there be an "else throw error" in here?